tools/src/main/java/org/apache/kafka/trogdor/workload/Throttle.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.trogdor.workload;
+
+import org.apache.kafka.common.utils.Time;
+
+import java.util.concurrent.TimeUnit;
+
+public class Throttle {
+    private final int maxPerSec;
+    private int count;
+    private long second;
+
+    Throttle(int maxPerSec) {
+        this.maxPerSec = maxPerSec;
+    }
+
+    synchronized public boolean increment() throws InterruptedException {
+        boolean throttled = false;
+        while (true) {
+            if (count < maxPerSec) {
+                count++;
+                return throttled;
+            }
+            long now = time().milliseconds();
+            long nextSecondMs = TimeUnit.MILLISECONDS.convert(second + 1, TimeUnit.SECONDS);
+            if (now < nextSecondMs) {
+                delay(nextSecondMs - now);
```

Comment:
When `maxPerSec` is high (say half of maximum throughput), doesn't this lead to overloaded half-second followed by an idle half-second, leading to high latencies? If this is only likely to be used at low throughput, a comment will be useful. Otherwise, using smaller delays with smaller intervals for the check may work better.

Reply:
That's a very good point, Rajini.  This could create artificially high latency when linger.ms > 0.

I came up with two fixes. Firstly, we now call Producer#flush before sleeping, in the throttle.  This should mark any pending messages as ready to be sent, even if the we haven't reached the batch size.  I was able to do this by making a subclass of Throttle.

Secondly, I reduced the period to 100 ms instead of 1000 ms, to make this a little bit less lumpy.  I didn't want to reduce the period too much, because the amount of time we delay when calling sleep() tends to vary a little bit on different platforms.  It's generally not that reliable when you get less than 10ms.